src/adaptations/device-layer/include/Weave/DeviceLayer/internal/FactoryProvisioning.ipp

```diff
@@ -0,0 +1,249 @@
+/*
+ *
+ *    Copyright (c) 2019 Google LLC.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Contains non-inline method definitions for the FactoryProvisioningBase
+ *          template.
+ */
+
+#ifndef FACTORY_PROVISIONING_IPP
+#define FACTORY_PROVISIONING_IPP
+
+#include <Weave/DeviceLayer/internal/WeaveDeviceLayerInternal.h>
+#include <Weave/DeviceLayer/internal/FactoryProvisioning.h>
+
+namespace nl {
+namespace Weave {
+namespace DeviceLayer {
+namespace Internal {
+
+// Fully instantiate the default FactoryProvisioningBase class.
+template class FactoryProvisioningBase<FactoryProvisioning>;
+
+template<class DerivedClass>
+WEAVE_ERROR FactoryProvisioningBase<DerivedClass>::ProvisionDeviceFromRAM(uint8_t * memRangeStart, uint8_t * memRangeEnd)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+    uint8_t * dataStart;
+    size_t dataLen;
+
+    // Search the given RAM region for device provisioning data.  If found...
+    if (Derived()->LocateProvisioningData(memRangeStart, memRangeEnd, dataStart, dataLen))
+    {
+        // Wipe the provisioning data marker so that it will not be found again should the device reboot.
+        memset(dataStart - FactoryProvisioningData::kMarkerLen, 0, FactoryProvisioningData::kMarkerLen);
+
+        // Parse the provisioning data TLV and write the values to persistent storage.
+        TLV::TLVReader reader;
+        reader.Init(dataStart, dataLen);
+        err = Derived()->StoreProvisioningData(reader);
+        SuccessOrExit(err);
+
+        // Wipe the provisioning data itself.
+        memset(dataStart, 0, dataLen);
```

Comment:
Any need to "securely" wipe the content?

Reply:
That seems reasonable.  Will fix.